### DISCUSSION TITLE
Matching the types inside type choice

### DISCUSSION THREAD

**Discussion by lsafina** [jolie/jolie#575] [Ideas (Features, Fixes, etc.)]
Dear all,

Since there is a possibility to create type choices (like for example `type animal: cat | dog`), probably some kind of matching directive could be implemented, like:

```
match:
  cat {//set of actions}
  dog {//another set of actions}
```

What do you think? Would it be useful?

**Comment by thesave** [comment]
That would be great :)

I remember having a conversation with @fmontesi about this. 
Even better if we had something similar to pattern matching à la Scala, e.g.,

```
expression match {
  case "\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b" => emailProcedure
  case cat => println@Console( "Found a cat with name" + cat.name )
  case _ => defaultProcedure
  }
```

 @jolie/developers, other opinions or use-cases on this?

**Comment by klag** [comment]
In Jolie there is the primitive instanceof for testing the match of a variable against a type.
it would be nice to extend such a primitive instead of introducing a new one

**Comment by thesave** [comment]
In the examples above `instanceof` could be used as condition for the cases e.g.,

``` java
expression match {
  case "\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b" => emailProcedure
  case instanceof cat => println@Console( "Found a cat with name" + cat.name )
  case _ => defaultProcedure
  }
```

However, I recognise that this might fall out of the concept of having a service-oriented language, where e.g., the matching of the regex (the first one) should be done with a call to a service (e.g., StringUtils) and not by some primitive of the language.

**Comment by lsafina** [comment]
We could use both match/case and instanceof, having match/case like syntactic sugar. 
I'm afraid it might be not always clear how to infer a type for pattern matching against the values, e.g is `\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b` regexp or just a string type?"
What if we start with matching against the types only?

By the way, should we do using of "case _" mandatory to guarantee matching of all possible types?

**Comment by thesave** [comment]
I second @klag in keeping the `instanceof` keyword for type introspection. 
I tried to put together a list of the most interesting (and applicable) cases for Jolie.

``` scala
selector match { alternatives }
```

Patters range over:
- constants, which check for equality

``` scala
 case 5                       => ...       // int
 case true                    => ...       // bool
 case "hello"                 => ...       // string
 case { .greeting = "hello",               
        .name = "world" }     => ...       // trees
```
- types, which check for type matching, with keyword `case instanceof`

``` scala
 case instanceof Type         => ...
```
- guarded, with keyword `case if` followed by a boolean expression and where `_`
  refers the selector

``` scala
 case if ( _ > 10 )           => ...
 case if ( _ && otherVar )    => ...
```
- regex matching, with keyword `case of` followed by a regular expression in string form 

``` scala
 case of "[a-zA-Z]+"           => ...
```
- wildcard, which matches any selector and refers it with keyword `_`. It might be used as default case

``` scala
 case _                        => ...
```

I tried to list the patterns in order of ease of development (although guarded, regex, and wildcard deserve some thoughts wrt variable bindings, etc.). As noted by @agwe, we could also include smart checks e.g., 

``` scala
match a {
 case _                        => ...
 case "hello"                  => ...
}
```

warning the user that the second case is unreachable.

**Comment by fmontesi** [comment]
Hi all,

I'm back in action. :-)

Checking a type choice can already be done using if-then-else and instanceof. For example, given a variable x of type `T1 | T2`, we can do:

```
if ( x instanceof T1 ) { ... } else if ( x instanceof T2 ) { ... }
```

So match must bring something new to the table. What is it, exactly? In some languages, match is an expression itself, i.e., I can write:

```
1 + x match { T1 => 2, T2 => 3 }  // evaluates to 3 if x has type T1, 4 if x has type T2
```

Can somebody clarify what's the intended use case for this?

Cheers,
Fabrizio

**Comment by fmontesi** [comment]
Hi all,

Any news on this?

Cheers,
Fabrizio

**Comment by thesave** [comment]
It's been quite some time :)

From what I recall my point was more on having some syntactic sugar to better structure code.

Tentatively, syntax 

```
e match {
  case [o_1] e_1 { ... }
  ...
  case [o_n] e_n { ... }
  case _ { ... }
}
```

means that expression `e` is matched against expressions `e_1`, ...,` e_n` where the matching operation is defined by the respective operators `o_1`, ..., `o_n`. When operator `o` is omitted we default to `==`. From top to bottom, the first matching expression executes the associated code within braces `{ ... }` and the other branches are discarded. Finally control bubbles up to the parent scope like we do for `if`s, `while`s (so we compose as usual `e match { ... }; ...`).

At runtime cases are translated into chains of `if`s. So, picking some examples wrote above (I excluded tree comparison and regexps, which probably deserve separate issues):

```Jolie
match e {
 case 5 { ... }  
 case true { ... }
 case "hello" { ... }
 case instanceof Type { ... }
 case > 10 { ... }
 case && e_2 { ... }
 case _ { ... }
}
```
we'd have

```Jolie
if ( e == 5 ){ ... }
else if ( e == true ) { ... }
else if ( e == "hello" ) { ... }
else if ( e instanceof Type ) { ... }
else if ( e > 10 ) { ... }
else if  ( e && e_2 ) { ... }
else { ... }
```
In matches `_` expands to `e`. Besides explaining why `case _` is the default case, `_` might come handing with chained evaluations on `e`, e.g., 

```
match e {
  case > 5 && _ < 10 { ... }
}
```

which translates to 

```
if ( e > 5 && e < 10 ) { ... }
```

Obviously the one above is just a thought and probably many @jolie/developers with more experience in developing production code can give some insight on interesting cases and the usefulness  and possible other interpretations of the construct :)
